### ISSUE TITLE
(5P) Podcast Player: Research how to work together with theme styles

### ISSUE THREAD

**Issue by marekhrabe** [Automattic/jetpack#15132]
Themes like 2020 break our layouts in editor, as they style extremely generic elements — `H2` or `OL` for example. 

Would some sort of CSS reset scoped for our player work around that? What else could work? We probably can't get around using `!important`s :D 

Themes should still be able to style out player as they wish but only using `.jetpack-podcast-player…` prefixed selectors and not things like generic `h2 { }` style definition

Example of style differences with Twenty Twenty:

| Editor | Frontend |
| - | - |
| <img width="610" alt="Screenshot 2020-03-26 at 09 27 08" src="https://user-images.githubusercontent.com/156676/77641956-2add2880-6f5d-11ea-80fb-76a796cf5667.png"> | <img width="587" alt="Screenshot 2020-03-26 at 09 26 40" src="https://user-images.githubusercontent.com/156676/77641973-316ba000-6f5d-11ea-8d50-0992e1657863.png">

Notice the episode list has numbered list style because 2020 declares it on any `ol`.

Another case with the episode title:

| Editor | Frontend |
| - | - |
| <img width="318" alt="Screenshot 2020-03-26 at 12 28 58" src="https://user-images.githubusercontent.com/156676/77642145-78599580-6f5d-11ea-8900-44bd03023ce9.png"> | <img width="306" alt="Screenshot 2020-03-26 at 12 29 09" src="https://user-images.githubusercontent.com/156676/77642151-7abbef80-6f5d-11ea-94fd-6f367a1156c4.png">

The `h2` headline (episode title) is misaligned with the image because the `h2` gets some added margin from the 2020 theme.

**Comment by frontdevde** [comment]
> Would some sort of CSS reset scoped for our player work around that? 

It could! That said if we were to reset theme styles we'd have to be very selective and careful as to which styles we'd override. Generally, I'd say we'd want most themes styles to apply. So we'd have to define what parts of the block we want to be very opinionated on e.g. do we think the list style is something we must necessarily control?

> We probably can't get around using !importants 

If we scope potential overrides to our block, then I'd hope we'd be able to make the selectors specific enough to override the theme styles without using !importants (unless the theme is using !important of course).

**Comment by jeryj** [comment]
I quickly tested out Varia, Maywood, and 2020. 2020 had some layout issues, yes, but there were pretty minimal overall, like one rule in the front-end and one rule on the editor layer.

Is there a default theme set that Jetpack supports? I can test with more themes to see if this issue would require more work. If it's just WPCOM themes, then it _seems_ like we may be OK without a theme reset.

**Comment by jeryj** [comment]
In general, the front-end styles need a little specificity boost to be applied correctly in the editor as well. How would y'all feel about just wrapping the entirety of the CSS in the `.wp-block-jetpack-podcast-player` class. So,

```
.wp-block-jetpack-podcast-player {
  // all of our styles here
  .jetpack-podcast-player__title {
    //
  }
  // etc.
}
```

This may be enough to get us to have a lot more consistency in our styles across different themes and the editor without being _too heavy handed_ about it. After all, we're dropping this inside of people's sites where we have no control over the code quality. It'd be nice to balance out expected visual styles and ability to be customized.

**Comment by WunderBart** [comment]
Just did a quick local check of what @jeryj suggested above and it nicely evened out the styles between editor & front-end. The only part different was the link outline, which is not currently addressed in Player's styles at all, so it just takes the default from the editor:

Scoped styles with the `.wp-block-jetpack-podcast-player` class:

| Editor  | Front-end |
| ------------- | ------------- |
| <img width="811" alt="Screenshot 2020-03-27 16 55 52" src="https://user-images.githubusercontent.com/1451471/77776806-de721580-704e-11ea-9e6f-77da90366d46.png"> | <img width="784" alt="Screenshot 2020-03-27 16 55 58" src="https://user-images.githubusercontent.com/1451471/77776812-e2059c80-704e-11ea-94c9-9e6ef7a2c613.png"> |

**Comment by marekhrabe** [comment]
That's some really good news! What about we get the style wrap merged now (as it clearly helps while having no downsides) and see how far we've got? It looks really promising from @WunderBart test screenshots.

We need to keep in mind that, as opposed to some of our previous work, this block in Jetpack works with much broader set of themes, compared to wpcom. I don't think doing per-theme fixes (however small) would be a good and scalable tactic so some sort of reset might need to be used. But let's not get too ahead and try the scoping first.

As for the outline, I will open a new issue. Let's get together with designers and come up with a cohesive `:focus` styles for our whole block (player + playlist).

**Comment by jeherve** [comment]
> We need to keep in mind that, as opposed to some of our previous work, this block in Jetpack works with much broader set of themes, compared to wpcom. I don't think doing per-theme fixes (however small) would be a good and scalable tactic so some sort of reset might need to be used. 

+1.

It's worth noting that in Jetpack, we really only do per-theme fixes for the default WordPress themes. We have a set of compatibility CSS files that are loaded for each default theme, and that can be edited to add such fixes:
https://github.com/Automattic/jetpack/tree/a49646f6be8862d8b58ff200f6ecb181cc312641/modules/theme-tools/compat

Those compat CSS files are synchronized with WordPress.com.

**Comment by WunderBart** [comment]
> What about we get the style wrap merged now (as it clearly helps while having no downsides) and see how far we've got? 

Ready for review here: https://github.com/Automattic/jetpack/pull/15201. It's branched off the https://github.com/Automattic/jetpack/pull/15211 because I was already finishing it up and thought I'd avoid some conflicts this way. https://github.com/Automattic/jetpack/pull/15211 is also ready BTW!

> As for the outline, I will open a new issue. Let's get together with designers and come up with a cohesive :focus styles for our whole block (player + playlist).

In https://github.com/Automattic/jetpack/pull/15201 I've just disabled the outline to make it consistent between editor and front-end for now: (done in https://github.com/Automattic/jetpack/pull/15201/commits/5602136d1266b4f2458181979f82e84432a7eeb3).

**Comment by marekhrabe** [comment]
@jeryj Is there a way to not use `H2` element for the heading but still give it a role of a heading with the same features as it has now? Being in the heading list, having the same audible feedback with a screen reader, etc?

I'm asking because just about any theme comes with opinionated styles for `H2` and we might not be able to account for everything, as some themes go really crazy. Like 2019 putting a `:before` pseudo element with decorative line above the heading etc… It's just very wide spectrum of things that can be applied to it. 

If we can use more generic element while keeping all the accessibility bells and whistles of a regular heading, I'd like to change it. If not, let's keep it as heading and try to add rules to our `H2` styles which will reset the most common styles — fonts, margins, paddings, maybe even hide pseudo elements and so on.

**Comment by jeryj** [comment]
You _can_ use `aria` roles to get the heading semantic, but it's definitely not encouraged. No idea on the screen reader support. MDN doesn't offer it: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/heading_role

I'm guessing any accessibility people we asked about this would tell us to definitely not use `aria` for this.

Would going with an `h3` make it any better? On the blog-roll index, the page titles all get an `h2`, which would mean we should use an `h3` in that context. But on an individual page with an `h1`, the podcast player should likely be an `h2`. I think you could argue that it _could_ be an `h3` though if that makes things a lot easier.

**Comment by obenland** [comment]
Let's optimize for the past few default themes and continue iterating after?

**Comment by obenland** [comment]
Fixed in #15320.
